[MCHECKSTYLE-381] remove call to checker.setClassLoader().

bmhm opened the pull request:
- See also: https://github.com/checkstyle/checkstyle/issues/7190

Signed-off-by: Benjamin Marwell <bmarwell@gmail.com>

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MCHECKSTYLE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MCHECKSTYLE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MCHECKSTYLE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

rnveach commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -201,8 +201,6 @@ public URLClassLoader run()
             }
         } );
 
-        checker.setClassLoader( projectClassLoader );
```
Variable `projectClassLoader` still remains and all it does it call close a few lines down. Seems to me it can be completely removed now.

bmhm replied:
Sure, sorry for the oversight.

bmhm commented:
The method `private List<File> getFilesToProcess( CheckstyleExecutorRequest request )` also has a unused variable `StringBuilder excludesStr = new StringBuilder();`. It is only updated but never read. Since my changes did not interfere with that part of the code, I left it the way it was.
Feel free to rework that part of code in another commit.

bmhm commented:
This plugin is using checkstyle 8.19 and will probably not update beyond 8.23, because 8.24 introduced breaking changes.
Not speaking for the maven development team, but if you remove the method within the next year or so, people might not be able to use this plugin with a somewhat recent version of checkstyle.

Maybe only mark it as deprecated and for removal in 2021 or so?

I'll file this on the Maven dev mailing list.

rfscholte commented:
This has already been mentioned before, but as long as checkstyle and the maven-checkstyle-plugin are 2 separate projects, it'll be hard to handle these API changes. It would be easier if the plugin would move to the checkstyle-project and maintained there (with or without help from someone of the Maven team).

romani commented:
checkstyle core library is very small team (3 maintainers) that collaborate at very limited free time.
We are not able to provide good feedback on Issues/PRs that we have from community.
We put under our umbrella plugins(eclipse, sonar) only when it is question of live of death, we have no time to properly support them. So their state under checkstyle organization is slightly better than death.
Without Maven team help we might not be able to provide good support for it, as Maven context is required.
We are open to provide some help with migration to new checkstyle versions for plugin, but release process and Maven nuances might be too big context for us to acquire.

bmhm commented:
It's really not easy to spot breaking changes from a maven perspective. When I recently updated checkstyle, my xml was suddenly invalid despite there was no major version update on behalf of checkstyle.

Maybe you might want to use semantic versioning and a more stable API with transition times (eg line length supported below tree walker and top level element) for some time.

That way the maven team might be able to support this plugin more easily. I'm writing from a user perspective here, and I wouldn't like this plugin to be near death. I am willing to create some more PR if needed.

struberg commented:
hi @romani We really appreciate all the time and passion you and the whole checkstyle team puts into it! Please don't interpret the feedback as criticism but as a discussion how we could improve long term stability together!
The Apache maven-checkstyle-plugin has a lot of integration tests which might e.g. give a quick feedback about backward compatibility. Probably your team could treat it as additional integration test coverage to run the maven-checkstyle-plugin with -Prun-its?
Wdyt, might strengthening the communication between the checkstyle and Maven teams improve the situation?

romani commented:
> Maybe you might want to use semantic versioning

It is not that easy in our almost 20 years old project.
Details:
https://checkstyle.org/#Development_Cycle
https://github.com/checkstyle/checkstyle/issues/3709

> and a more stable API with transition times

API is too big, bunch of very very old decisions based on lack of time many years ago ..... .
We do try to break API one by one. This API change is so long awaited, I started to warn about its removal probably about 5 years ago. But till it stop to compile or work .... nobody care.

> Please don't interpret the feedback as criticism

thanks for confirmation, and I do not misinterpretate it, It completely natural for plugins to live close to core library, and it will be much better for plugin .... but we are not ready to change playground of project .... if it goes forward ... let it go for now as it is. 

> Probably your team could treat it as additional integration test coverage to run the maven-checkstyle-plugin with -Prun-its?

we do not run this ... BUT we heavily depend on maven plugin to keep functional regression (huge amount of execution of plugin in different projects) on leash. If we are introducing breaking changes for plugin, we get to know this from CI when PR is created to checkstyle project.
That is why we would like some changes to be more released more frequently as we can not remove code till maven plugin is migrated off it.

> might strengthening the communication between the checkstyle and Maven teams improve the situation?

We are always open to help, just mention us in PR and we will help to resolve migrations to new version of checkstyle or any other core library questions.

bmhm commented:
@romani @struberg @rfscholte 

Reworked to call the method if it is available by catching a `NoSuchMethodError`. I was also thinking of using reflection to check if the method exists, but I decided agains this to have minimal changes.
However, I changed the `urls.toArray()` method to use `new URL[0]` which is faster than `new URL[urls.size()]` and also the recommended way on more recent versions (Java 8+).

Please review.

bmhm commented:
Thanks and agreed about the tests. We may need to refibe existing tests if/when we use checkstyle 8.24 or newer.

bmhm commented:
For users? No. This is really interesting for devs. Imagine you do not upgrade checkstyle (yet) for various reasons, but the checkstyle plugin. You will start seeing warnings about some internal unused methods. I do not believe they are useful to maven users.

If you think they are, I'd love to hear why to implement that warning. 
Please don't take this as a criticism. I just want to avoid warnings issued to users which are not that important to them as they may seem helpful to us.

romani commented:
> You will start seeing warnings about some internal unused methods.

it is better than live in illusions that classpath is actually helping chekstyle to do validation. Classpath functionality was completely removed in 8.28 (released today), it was very unreliable in validation from the beginning. So users unlikely happy with classpath related features.

> I'd love to hear why to implement that warning.

I am not ready to provide good reasons. If you not feel that it is good - I trust you. Lets proceed without warning.
 I just try to send a message that any attempts to provide classpath to plugin will be ignored by library/validator. I thought in some discussion that you are in favor of deprecation process, I agree with you to some extend, we need to give signal to user that classpath is deprecated (Checks do not use classpath).

bmhm commented:
> we need to give signal to user that classpath is deprecated

Yes I agree on that. That's definitely important. Maybe add the warning in a later version of checkstyle maven plugin?

The no op from checkstyle change is quite recent. If it wasn't, I had less trouble for such a message to alert the user to upgrade.

But this plugin is using checkstyle 8.19. Upgrading beyond 8.23 is a breaking change due to changes in the checkstyle.xml layout.

What do you think? I'll file another PR in an hour or so for the version after the next version.

romani commented:
I am ok, thanks a lot.

eolivelli commented:
waiting for CI
https://builds.apache.org/job/maven-box/job/maven-checkstyle-plugin/job/MCHECKSTYLE-381/

Will commit after a blue light

bmhm commented:
@eolivelli blue light spotted :)
https://builds.apache.org/job/maven-box/job/maven-checkstyle-plugin/job/MCHECKSTYLE-381/2/

eolivelli commented:
CI is blue

eolivelli commented:
committed as eae07f99f01584bfd3da90a8c5eb32364e8ee82b

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
         return checkerListener.getResults();
     }
 
+    private void setUpCheckstyleClassloader( Checker checker,
```
nit: ClassLoader

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
+                                             List<String> outputDirectories )
+        throws CheckstyleExecutorException
+    {
+        final List<URL> urls = new ArrayList<>( classPathStrings.size() );
```
personal preference; but I tend not to use final on local variables where it's not required. Up to you though.

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
+        {
+            try
+            {
+                urls.add( new File( path ).toURI().toURL() );
```
Paths.get(path).toURI() would also work. I'm not sure if this matters in any way.

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
+                if ( outputDirectoryString != null )
+                {
+                    File outputDirectoryFile = new File( outputDirectoryString );
+                    if ( outputDirectoryFile.exists() )
```
I think you want isDirectory here instead of exists

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
+                    if ( outputDirectoryFile.exists() )
+                    {
+                        URL outputDirectoryUrl = outputDirectoryFile.toURI().toURL();
+                        getLogger().debug( "Adding the outputDirectory " + outputDirectoryUrl.toString()
```
not sure this log message carries its weight

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
+            }
+            catch ( MalformedURLException e )
+            {
+                throw new CheckstyleExecutorException( e.getMessage(), e );
```
Check if this is what you want. It might be, but just to be sure, you want a single bad URL to fail the method even if the other URLs are OK, and before processing the other URLs?

elharo commented on src/main/java/org/apache/maven/plugins/checkstyle/exec/DefaultCheckstyleExecutor.java:
```diff
@@ -316,6 +258,73 @@ public URLClassLoader run()
+
+        URLClassLoader projectClassLoader = AccessController.doPrivileged( new PrivilegedAction<URLClassLoader>()
+        {
+            public URLClassLoader run()
```
`@Override`

also, this would be clearer as a non-anonymous class or even an outer class

bmhm commented:
@elharo  this is already merged and on maven central.
Also, everything you mentioned will be removed in the version after the next one.
